x-pack/metricbeat/module/aws/ec2/ec2.go

```diff
@@ -0,0 +1,313 @@
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet(aws.ModuleName, "ec2", New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MockEC2Client struct is used for unit tests.
+type MockEC2Client struct {
+	ec2iface.EC2API
+}
+
+// MockCloudWatchClient struct is used for unit tests.
+type MockCloudWatchClient struct {
+	cloudwatchiface.CloudWatchAPI
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*aws.MetricSet
+	config *aws.Config
+}
+
+// metricIDNameMap is a translating map between createMetricDataQuery id
+// and aws ec2 module metric name.
+var metricIDNameMap = map[string]string{
+	"cpu1":     "cpu.total.pct",
+	"cpu2":     "cpu.credit_usage",
+	"cpu3":     "cpu.credit_balance",
+	"cpu4":     "cpu.surplus_credit_balance",
+	"cpu5":     "cpu.surplus_credits_charged",
+	"network1": "network.in.packets",
+	"network2": "network.out.packets",
+	"network3": "network.in.bytes",
+	"network4": "network.out.bytes",
+	"disk1":    "diskio.read.bytes",
+	"disk2":    "diskio.write.bytes",
+	"disk3":    "diskio.read.ops",
+	"disk4":    "diskio.write.ops",
+	"status1":  "status.check_failed",
+	"status2":  "status.check_failed_system",
+	"status3":  "status.check_failed_instance",
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The aws ec2 metricset is experimental.")
+
+	config := aws.Config{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+		config:    &config,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	//mock Fetch
+	if m.config.AwsConfigProfileName == "" {
+		m.MockFetch(report)
+		return
+	}
+
+	//actual fetch function
+	cfg, err := external.LoadDefaultAWSConfig(external.WithSharedConfigProfile(m.config.AwsConfigProfileName))
+	if err != nil {
+		logp.Err("Failed to load config: %s", err.Error())
+	}
+
+	cfg.Region = "us-west-1"
+	svcEC2 := ec2.New(cfg)
+	//Get a list of regions
+	regionsList, err := getRegions(svcEC2)
+	if err != nil {
+		report.Error(errors.Wrap(err, "getRegions failed"))
```

Comment:
I would add the return here too, although this one should be safe, it removes uncertainty